File: ios/MoodleAssessmentApp/MoodleAssessmentApp/Data/Mapping/CourseMapper.swift
import Foundation

enum CourseMapper {
    static func map(_ dto: CourseDTO) -> Course {
        let title = dto.displayname ?? dto.fullname ?? "Untitled Course"
        let progress = dto.progress ?? dto.completionprogress
        let imageURL =
            dto.overviewfiles?.first.flatMap { URL(string: $0.fileurl) } ??
            dto.courseimage.flatMap(URL.init(string:))

        return Course(
            id: dto.id,
            name: title,
            progress: progress,
            imageURL: imageURL
        )
    }
}

File: android/app/src/main/java/com/itcorner/moodleandroid/data/mappers/CourseMapper.kt
package com.itcorner.moodleandroid.data.mappers

import com.itcorner.moodleandroid.data.dto.CourseDto
import com.itcorner.moodleandroid.domain.Course

object CourseMapper {
    fun map(dto: CourseDto): Course {
        val preferredName = dto.displayname?.takeIf { it.isNotBlank() }
            ?: dto.fullname?.takeIf { it.isNotBlank() }
            ?: "Untitled course"

        val resolvedImage = dto.courseimage?.takeIf { it.isNotBlank() }
            ?: dto.overviewfiles?.firstOrNull { !it.fileurl.isNullOrBlank() }?.fileurl

        return Course(
            id = dto.id,
            name = preferredName,
            progress = dto.progress ?: dto.completionprogress,
            imageUrl = resolvedImage
        )
    }
}
